Provide an API for applications to be able to transform the return values from their `source` function before rendering the suggestions.

JakeChampion opened the pull request:
The use-case that this is required for, is when the `source` function is returning data that is not an array of strings, such as an array of objects. For example the objects could contain an ID field and a string field so that they can more easily track what suggestion was selected.

JakeChampion committed:
Provide an API for applications to be able to transform the return values from their `source` function before rendering the suggestions.

The use-case that this is required for, is when the `source` function is returning data that is not an array of strings, such as an array of objects. For example the objects could contain an ID field and a string field so that they can more easily track what suggestion was selected.

JakeChampion committed:
rename type from Suggestion to Source

JakeChampion committed:
Add custom type for the optionToSuggestion function

JakeChampion committed:
Apply the optionToSuggestion function when a suggested is selected, to ensure that the input's value contains the correct string

JakeChampion committed:
Add a hidden demo for manually testing out the optionToSuggestion API of the component

JakeChampion committed:
Add test to confirm optionToSuggestion works as expected

JakeChampion committed:
Change type on PopulateResults to any because the application can now defined any return type, as long as they have a `optionToSuggestion` function which converts the results to strings

JakeChampion commented on src/js/autocomplete.js:
```diff
@@ -135,21 +135,27 @@ function initClearButton(instance) {
  * @typedef {Object} AutocompleteOptions
- * @property {Suggestions} source - The function which retrieves the suggestions to display
+ * @property {Source} source - The function which retrieves the suggestions to display
+ * @property {Function} [optionToSuggestion] - Function which transforms a suggestion before rendering
```
I'm not certain that the function name `optionToSuggestion` is a good name, please let me know of any suggestions for a different name you may have

notlee replied:
I think `optionToSuggestedValue` is more descriptive. A "suggested value" makes sense as the return of `templates.inputValue` and is also the name of the argument for `this.suggestionTemplate`

notlee replied:
Maybe even `mapOptionToSuggestedValue`

JakeChampion replied:
I like `mapOptionToSuggestedValue` -- I've updated the code to use that name in [`30c0c14` (#12)](https://github.com/Financial-Times/o-autocomplete/pull/12/commits/30c0c1411c73dae2c5a2dee98103ff5f299804ef)

notlee commented on src/js/autocomplete.js:
```diff
@@ -167,7 +173,44 @@ class Autocomplete {
+					// Apply the function to the option. This is a way for the
+					// consuming application to decide what text should be
+					// shown for this option.
+					// This is usually defined when the option is not already a string.
```
Should we error if `option` is not a string and `optionToSuggestion` is not set?

JakeChampion replied:
Good suggestion 👍 
I've added that in this commit [`c78e6ed` (#12)](https://github.com/Financial-Times/o-autocomplete/pull/12/commits/c78e6edec54b177b2cb8393a6cf4edfb688ef88f)
It was hard to test due to the fact that accessible-autocomplete uses preact to render and preact renders asynchronously, but, it is possible to test :D

notlee commented:
This looks good. :) I like that it's giving the component the flexibility currently needed without letting`templates.suggestion` be fully customised. Could you update the readme too with an explanation of the usecase and a nice example?

JakeChampion committed:
rename optionToSuggestion to mapOptionToSuggestedValue

JakeChampion committed:
throw an error if the option attempting to be displayed in not a string and the mapOptionToSuggestedValue function is not defined

JakeChampion committed:
use the word options instead of results to keep consistency within the documentation

It is now possible for options to be any type, not only a string. Let's use `option` when talking about the values returned by the supplied `source` function and let's use `suggestion` when talking about the string which is displayed within the suggestions menu and into the autocomplete input box

JakeChampion committed:
Add documentation and example for mapOptionToSuggestedValue to readme

JakeChampion commented:
@notlee There is now an example and documentation within the readme 👍

JakeChampion committed:
change test setup to enable the onunhandledrejection assertion to work when running karma in ci

notlee commented on README.md:
```diff
@@ -146,7 +146,49 @@ If wanting to supply dynamic suggestions, you will need to provide a function wh
+
+### mapOptionToSuggestedValue
+
+This function is used to convert the options returned from the `source` function into a their respective strings to be used as the suggestion values within the suggestions menu.
```
Started this suggestion to correct the "into a their" typo, but think it can be simplified a bit, what do you think?
```suggestion
This function is used to convert the options returned from the `source` function into strings for the suggestions menu.
```

JakeChampion replied:
I like it 👍

JakeChampion committed:
Simplify the description of mapOptionToSuggestedValue

Co-authored-by: Lee Moody <notlee@users.noreply.github.com>

JakeChampion committed:
Add mapOptionToSuggestedValue to this.options

origamiserviceuser commented:
:tada: This PR is included in version v1.1.0 :tada:
